opm/parser/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -766,6 +772,32 @@ namespace Opm {
 
         return index(i,j,k,c);
     }
+
+    size_t ZcornMapper::fixupZCORN( std::vector<double>& zcorn, int sign, size_t i, size_t j, size_t k , size_t c) {
+        size_t index1 = this->index(i,j,k,c);
+        size_t index2 = this->index(i,j,k,c+4);
+
+        if ((zcorn[index2] - zcorn[index1]) * sign < 0 ) {
+            zcorn[index2] = zcorn[index1];
+            return 1;
+        } else
+            return 0;
+
+    }
+
+
+    size_t ZcornMapper::fixupZCORN( std::vector<double>& zcorn) {
+        int sign = zcorn[ this->index(0,0,0,0) ] < zcorn[this->index(0,0, this->dims[2] - 1,0)] ? 1 : -1;
```

Comment:
This is the bug. It looks at the top and bottom cells' top coordinates, but if there is only one layer this is the same cell, and sign becomes -1.

To fix it I tested with replacing the final 0 with a 4 (comparing to the bottom cell), that should fix all trivial cartesian cases, although for a corner-point grid there is no requirement that this should work.

However if we also replace the < with a <= I think we will probably never encounter a problematic grid, unless designed to beat this test! In summary, I suggest the line:

```
        int sign = zcorn[ this->index(0,0,0,0) ] <= zcorn[this->index(0,0, this->dims[2] - 1,4)] ? 1 : -1;
```